File: src/treemapper/diffctx/parsers/generic.py
from __future__ import annotations

from pathlib import Path

from ..stopwords import TokenProfile
from ..types import Fragment, FragmentId, extract_identifiers
from .base import GENERIC_MAX_LINES, find_smart_split_point


class GenericStrategy:
    priority = 0

    def can_handle(self, _path: Path, _content: str) -> bool:
        return True

    def fragment(self, path: Path, content: str) -> list[Fragment]:
        lines = content.splitlines()
        if not lines:
            return []

        fragments: list[Fragment] = []
        total = len(lines)
        start_idx = 0

        while start_idx < total:
            target_end_idx = min(total - 1, start_idx + GENERIC_MAX_LINES - 1)
            end_idx = find_smart_split_point(lines, start_idx, target_end_idx, path)

            snippet = "\n".join(lines[start_idx : end_idx + 1])
            if not snippet.endswith("\n"):
                snippet += "\n"

            profile = TokenProfile.from_path(str(path))

            fragments.append(
                Fragment(
                    id=FragmentId(path=path, start_line=start_idx + 1, end_line=end_idx + 1),
                    kind="chunk",
                    content=snippet,
                    identifiers=extract_identifiers(snippet, profile=profile),
                )
            )
            start_idx = end_idx + 1

        return fragments

File: diffctx/src/parsers/generic.rs
use std::sync::Arc;

use crate::config::parsers::PARSERS;
use crate::config::tokenization::TOKENIZATION;
use crate::types::{Fragment, FragmentId, FragmentKind, extract_identifiers};

use super::FragmentationStrategy;

pub struct GenericStrategy;

impl FragmentationStrategy for GenericStrategy {
    fn can_handle(&self, _path: &str, _content: &str) -> bool {
        true
    }

    fn fragment(&self, path: Arc<str>, content: &str) -> Vec<Fragment> {
        let lines: Vec<&str> = content.split('\n').collect();
        if lines.is_empty() {
            return Vec::new();
        }

        let total = lines.len();
        let mut fragments: Vec<Fragment> = Vec::new();
        let mut start_idx: usize = 0;

        while start_idx < total {
            let end_idx = (start_idx + PARSERS.generic_max_lines - 1).min(total - 1);

            let start_line = start_idx as u32 + 1;
            let end_line = end_idx as u32 + 1;

            let mut snippet = lines[start_idx..=end_idx].join("\n");
            if !snippet.ends_with('\n') {
                snippet.push('\n');
            }
            let identifiers =
                extract_identifiers(&snippet, TOKENIZATION.fragment_min_identifier_length);

            fragments.push(Fragment {
                id: FragmentId::new(Arc::clone(&path), start_line, end_line),
                kind: FragmentKind::Chunk,
                content: Arc::from(snippet),
                identifiers,
                token_count: 0,
                symbol_name: None,
            });

            start_idx = end_idx + 1;
        }

        fragments
    }
}
